### ISSUE TITLE
Why do we believe it's a good idea to return an issued credential to an internal caller?

### ISSUE THREAD

**Issue by dhh1128** [w3c-ccg/vc-http-api#48]
Issued credentials have to flow to *holders*. This API requires issuers to also support flowing them to an internal caller for some reason. I am not aware of any architectural reason that's rooted in use cases, why this should be so. The protocol for issuance that's imagined in [Aries RFC 0036](https://github.com/hyperledger/aries-rfcs/blob/master/features/0036-issue-credential/README.md) does not preclude sending an issued credential to an internal destination, but neither does it require that.

**Comment by bumblefudge** [comment]
i will try to formulate something

**Comment by msporny** [comment]
We discussed this on the 2021-08-03 call and @dmitrizagidulin and @bumblefudge agreed to write up the discussion in the group in this issue and then see if @dhh1128 agrees with the path forward.

**Comment by msporny** [comment]
Status on this, @dmitrizagidulin and @bumblefudge?

**Comment by msporny** [comment]
We discussed this on the 2022-03-29 call. @dmitrizagidulin summarized that the issue is about internal vs. external API -- who is going to use this? Back-end only APIs, Wallet APIs? Microcosm of entire spec, no actionable item yet. Conversation has progressed since initial API. Protect every endpoint with authz, shifts backend vs. public conversation a bit. @dlongley said we could close issue by saying "internal caller might be entity that passes VC to holder". @dmitrizagidulin said that's how DCC has VC API implemented. Wallet requests credential from Issuer, Issuer calles this API endpoint, then returns it back to wallet.

**Comment by tplooker** [comment]
@msporny I see this is marked as pending closed, does that mean the decision is to treat the VC API as an internal style API in the context of flows like the DCC usecase? E.g the endpoint definition that the wallet interacts with in that case is out of scope for this specification?

**Comment by msporny** [comment]
> @msporny I see this is marked as pending closed, does that mean the decision is to treat the VC API as an internal style API in the context of flows like the DCC usecase?

I added a "Summary" above as the notes from the call were pretty terrible.

In general, the VC API covers every action that you need for basic interaction within the Issuer/Holder/Verifier ecosystem. It covers all APIs necessary to issue a VC from an Issuer -> Holder. It covers all APIs necessary for a Holder to manage their VCs. It covers all APIs necessary for a presentation between a Holder -> Verifier. Some of these APIs, like /credentials/issue, which is typically authz'd by OAuth2 or ZCAPs and exposed on the Issuer Service (an "internal" system component), is expected to be called by the Issuer App (an "external" system component, exposed to the Internet). However, some system configurations, like SaaS, might expose /credentials/issue directly via the Issuer App and still do authz via OAuth2 or ZCAPs.

>  E.g the endpoint definition that the wallet interacts with in that case is out of scope for this specification?

It depends on what part of the wallet we're talking about. I know that the Traceability folks are using the VC API as their "wallet interface" of sorts. I know that Digital Bazaar is using the exchanges endpoints as the mechanism to move VPs from Issuer -> Holder or from Holder -> Verifier when CHAPI isn't utilized (like when doing credential refresh). Digital Bazaar, however, is using EDVs as the interface that the Holder uses to store/retrieve their VCs and is then using VC API as the mechanism to create presentations from those VCs (we aren't using the VC API credential/presentation storage/retrieval endpoints to do that like the Traceability folks are doing). 

Does that answer your question?

**Comment by tplooker** [comment]
> Does that answer your question?

Sort of, bear with me :).

I still don't seem to see how an e2e flow would work in the case of say issuing a personal identity credential (e.g DL or PRC). To me this flow involves three entities, the wallet, issuer and End-User. Typically unless you are assuming the wallet is handed some token of authorization that identifies who the End-User is and proves the wallet has the authority to request the credential from the issuer, then this has to be done within the protocol flow. I understand that designing of authorization protocols is out of scope for this spec and that its relationship is optional, but how do they relate when they need to?

Say you are using plain OAuth2 or OIDC to do user auth and issue an access_token, does the wallet include this somehow in a request to the interact API? Does that make the VC API a resource server of the IDP? What scopes need to be granted to get this access_token or does it not matter?

Maybe im missing the point, I'm just struggling to see how the protocol flow works e2e for that case?

**Comment by msporny** [comment]
> I still don't seem to see how an e2e flow would work in the case of say issuing a personal identity credential (e.g DL or PRC).

Yes, this is one of those discussions that we haven't been able to get to the heart of. At the center of the discussion is the question of how an Issuer/Verifier ensures the proper level of authz before engaging with a Holder.

One approach is to place some level of trust in the IdP/Holder software authz mechanism to the Issuer/Verifier. The OAuth2/OIDC approach is where you register a software client with the Issuer/Verifier. I'll note that many of the "internal" VC API endpoints use OAuth2 (or ZCAPs) for software client authz.

Another approach is to state there there should not be trust in the IdP/Holder software authz mechanism for two reasons 1) because it doesn't protect against particular types of attacks, and 2) it has negative effects on market competition in the ecosystem. The former could be argued to be more important than the latter.

Rather than have that discussion in this issue, which is related (but tangentially), I've raised a new issue (#279) to explore why the /exchanges endpoint [does authn/authz of the Holder in the message exchange](https://w3c-ccg.github.io/vc-api/#exchange-examples) rather than solely depending on OAuth2 on the API endpoint to perform the authz.

**Comment by msporny** [comment]
The group discussed this on the 2023-04-11 VC API telecon:

@jandrieu noted that we have had a lot of discussion around the difference between an "Issuer Coordinator" and "Issuer Service" and have moved away from the language related to "internal vs. external". We have also ensured that implementation of any of these mechanisms is optional and have provided one way in which interoperability can be achieved (e.g. implement the "issuer service API" or implement the "issuer coordinator API"). It was also noted that we don't have the concept of a "wallet" and those boundaries are less well defined... wrt. OAuth2  -- we haven't talked about how OAuth2 and OIDC are supported, except that exchanges have been demonstrated to support that. 

If we were to document how VC API works w/ OAuth2 or OID4, then we might address the question from @tplooker.

@dlongley noted wrt. making OID4 work w/ exchanges, Digital Bazaar has already implemented that, perhaps this is a matter of documenting how OID4 can be implemented with exchanges and that would allow us to close this issue out.

@jandrieu noted that a PR for this might require implementation guidance... if you want to integrate OID4 w/ VC API Exchanges, a section on how you do that would answer Tobias' question in terms of flow. We might need Exchangers to be integrated, might be in an example of how you use exchanges. An exchanger is probably a first class component that has its own endpoints and we haven't worked through that stuff yet.

The next step here is to raise a PR that details how OID4 can be integrated w/ VC API Exchanges.
